app/questionnaire/relationship_router.py

```diff
@@ -2,54 +2,126 @@
-            block_id=block_id,
-            list_item_ids=list_item_ids,
-            list_name=list_name,
-        )
+    UNRELATED_RELATIONSHIP_VALUE = "Unrelated"
+    UNRELATED_NO_ANSWER_VALUE = "No"
+
+    def __init__(
+        self,
+        answer_store,
+        relationship_store,
+        section_id,
+        list_name,
+        list_item_ids,
+        relationships_block_id,
+        unrelated_block_id=None,
+        unrelated_answer_id=None,
+    ):
+        self.answer_store = answer_store
+        self.relationship_store = relationship_store
+        self.section_id = section_id
+        self.list_name = list_name
+        self.list_item_ids = list_item_ids
+        self.relationships_block_id = relationships_block_id
+        self.unrelated_block_id = unrelated_block_id
+        self.unrelated_answer_id = unrelated_answer_id
+        self.path = self._generate_relationships_routing_path()
 
     def can_access_location(self, location):
         return location in self.path
 
-    def get_first_location_url(self, resume=False):
-        if resume:
-            return self.path[0].url(resume=resume)
-        return self.path[0].url()
+    def get_first_location(self):
+        return self.path[0]
 
-    def get_last_location_url(self):
-        return self.path[-1].url()
+    def get_last_location(self):
+        return self.path[-1]
 
-    def get_next_location_url(self, location):
+    def get_next_location(self, location):
         try:
             location_index = self.path.index(location)
-            return self.path[location_index + 1].url()
+            return self.path[location_index + 1]
         except IndexError:
             return None
 
-    def get_previous_location_url(self, location):
+    def get_previous_location(self, location):
         location_index = self.path.index(location)
         if not location_index:
             return None
-        return self.path[location_index - 1].url()
+        return self.path[location_index - 1]
+
+    def get_remaining_relationships_for_individual(self, current_location):
+        path_for_individual = [
+            location
+            for location in self.path
+            if location.list_item_id == current_location.list_item_id
+        ]
+        current_index = path_for_individual.index(current_location)
+        return [
+            location.to_list_item_id
+            for location in path_for_individual[current_index + 1 :]
+        ]
 
-    @staticmethod
-    def _generate_relationships_routing_path(
-        section_id, block_id, list_item_ids, list_name
+    def _generate_relationships_routing_path(self):
+        path = []
+        for from_item in self.list_item_ids:
+            from_index = self.list_item_ids.index(from_item)
+            path += self._get_individual_relationships_routing_path(
+                from_list_item_id=self.list_item_ids[from_index],
+                to_list_item_ids=self.list_item_ids[from_index + 1 :],
+            )
```

Comment:
Updated to use enumerate. I have also renamed `from_item` to `from_item_id` for clarity and replaced `self.list_item_ids[from_index]` with `from_list_item_id`.

You have also suggested to use `extend` rather than `+=`, is there a reason for that? I preferred `+=` as it avoided nesting brackets and was easier to read.

Reply:
No specify reason for `extend`, it is just what I normally use but more than happy for it to stay as `+=` 👍